oauth2_http/java/com/google/auth/oauth2/S2A.java

```diff
@@ -0,0 +1,110 @@
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ServiceLoader;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utilities to fetch the S2A (Secure Session Agent) address from the mTLS configuration.
+ *
+ * <p>mTLS configuration is queried from the MDS MTLS Autoconfiguration endpoint.
+ */
+@ThreadSafe
+public final class S2A {
+  public static final String S2A_CONFIG_ENDPOINT_POSTFIX =
+      "/computeMetadata/v1/instance/platform-security/auto-mtls-configuration";
+
+  public static final String METADATA_FLAVOR = "Metadata-Flavor";
+  public static final String GOOGLE = "Google";
+  private static final int MAX_MDS_PING_TRIES = 3;
+  private static final String PARSE_ERROR_S2A = "Error parsing S2A Config from MDS JSON response.";
+
+  private S2AConfig config;
+
+  private transient HttpTransportFactory transportFactory;
+
+  public S2A() {}
+
+  public void setHttpTransportFactory(HttpTransportFactory tf) {
+    this.transportFactory = tf;
+    this.config = getS2AConfigFromMDS();
+  }
+
+  /** @return the mTLS S2A Address from the mTLS config. */
+  public String getMtlsS2AAddress() {
+    return config.getMtlsAddress();
+  }
+
+  /** @return the plaintext S2A Address from the mTLS config. */
+  public String getPlaintextS2AAddress() {
+    return config.getPlaintextAddress();
+  }
+
+  /**
+   * Queries the MDS mTLS Autoconfiguration endpoint and returns the {@link S2AConfig}.
+   *
+   * <p>Returns {@link S2AConfig} with empty addresses on error.
+   *
+   * @return the {@link S2AConfig}.
+   */
+  private S2AConfig getS2AConfigFromMDS() {
+    String url = getMdsMtlsEndpoint();
+    GenericUrl genericUrl = new GenericUrl(url);
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY); 
+    if (transportFactory == null) {
+      transportFactory =
+          Iterables.getFirst(
+              ServiceLoader.load(HttpTransportFactory.class), OAuth2Utils.HTTP_TRANSPORT_FACTORY);
+    }
+
+    HttpRequest request;
+    try {
+      request =
+              transportFactory.create().createRequestFactory().buildGetRequest(genericUrl);
+      request.setParser(parser);
+      request.getHeaders().set(METADATA_FLAVOR, GOOGLE);
+      request.setThrowExceptionOnExecuteError(false);
+    } catch (IOException e) {
+      return S2AConfig.createBuilder().build();
+    }
+
+    for (int i = 0; i < MAX_MDS_PING_TRIES; i++) {
```

Comment:
I am not super familiar with google-http-client, but can this retry logic be handled by maybe `HttpRequestRetryHandler`?

`ExponentialBackOff` used in retrys [here](https://github.com/googleapis/google-auth-library-java/blob/d4a3763799896a35e27b9f5c178f77364e41668d/oauth2_http/java/com/google/auth/oauth2/IamUtils.java#L130C24-L143) is not thread-safe.
cc.@lqiu96 who might be more familiar

Reply:
Thanks for pointing this out Min! A few followup questions: 
- I'm not using `ExponentialBackOff` in the retries, is their still a thread-safety problem? The retry logic here is mirrored from https://github.com/googleapis/google-auth-library-java/blob/dde6876154d6e7ff43dbbf7b75c5205eadd679a5/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L435
- From quickly looking at https://cloud.google.com/java/docs/reference/google-http-client/latest/com.google.api.client.http, I don't see anything similar to `HttpRequestRetryHandler`. Should we instead use the `org.apache.http.impl.client` implementation so we can use `org.apache.http.impl.client.DefaultHttpRequestRetryHandler` or something similar?

Reply:
I only mentioned `ExponentialBackOff` because that is used in one of the recent changes related to retry (https://github.com/googleapis/google-auth-library-java/pull/1452/files)
I suppose you can do similar, by using HttpRequest's built-in retry handlers and set  `request.setNumberOfRetries(OAuth2Utils.DEFAULT_NUMBER_OF_RETRIES);`?  If you do not need `ExponentialBackOff`, maybe something simple e.g.

```
      // Set the number of retries
      request.setNumberOfRetries(OAuth2Utils.DEFAULT_NUMBER_OF_RETRIES); 
      // Retry on specific status codes (you might want to adjust these)
      request.setUnsuccessfulResponseHandler(
          new HttpUnsuccessfulResponseHandler() {
            @Override
            public boolean handleResponse(
                HttpRequest request, HttpResponse response, boolean supportsRetry)
                throws IOException   {
              return   RETRYABLE_STATUS_CODES.contains(response.getStatusCode());
            }
          });
```
About thread-safety, noticed HttpRequest is also marked as not safe [here](https://github.com/googleapis/google-http-java-client/blob/4830ad788a62fe9cd4f64873b771e6ef8ef92193/google-http-client/src/main/java/com/google/api/client/http/HttpRequest.java#L44), likely because its member variable hold state information . But I don't see it a concern in this implementation, as `executeAsync()` is not used and `getS2AConfigFromMDS()` creates and executes a request each time.

Reply:
Yeah I agree and I would prefer if the loops could be modified to use the unsuccessfulResponseHandler and exponential backoff (configured with a list of status codes that can be retried on).

Reply:
Thanks for the detail @zhumin8 ! In https://github.com/googleapis/google-auth-library-java/pull/1400/commits/12b248d8245162d63a1fe4863921b4cb595afa48 I have modified to set the number of retries on the request and use `ExponentialBackOff` and `HttpUnsuccessfulResponseHandler`. We only want to be retrying on 5xx error codes, so I have specified them as retryable error codes.